8366232: JFR startup messages are shown with -Xlog:jfr+startup=warning

parttimenerd opened the pull request:
Only print the JFR startup messages when no or a < warning log level is set explicitly by the user.

dholmes-ora commented:
/label add hotspot-jfr
/label add hotspot-runtime
/label remove hotspot

dholmes-ora reviewed (commented):
I can't say I like the general changes to UL to accommodate this. I also don't really like what JFR is already doing. I'm trying to see what the motivation is here. It appears to be "give a verbose startup message but don't label it as a 'warning'".  But there is also no guarantee that `jfr+startup` is being logged to `stdout` so this could by-pass the user's control over logging output!

parttimenerd commented:
The motivation is to better capture the user's intent. When the user doesn't set any warning level, then we keep the default behaviour, for backwards compatibility as in the original change, when the user sets the logging level, then we consider it.

You mentioned yourself in the original issue:

```
 I'm just not sure we have a way to query if the logging level has been explicitly set to a given value.
```

I just implemented this. I just implemented a way to properly check whether the user didn't specify a logging level, so the code doesn't have to assume that a `warning` logging level means the user didn't set anything.

This change tries to reduce confusion.

parttimenerd commented:
But isn't my PR closer to what the user expects? When you explicitly set the level to `warning`, no `info` level messages should be printed

dholmes-ora commented:
Functionally yes, but I'm just not sure about the mechanism. Need to think more about it. This reminds me of the flag complexity: FLAG_IS_DEFAULT, FLAG_IS_CMDLINE etc.

parttimenerd commented:
Regarding files:

`-Xlog:jfr+startup=info:file=a.txt` causes the startup message to be written to the file, but `-Xlog:jfr+startup=warning:file=a.txt` doesn't.

So it works both for files and stdout, the implementation changes the behaviour at the root.
